python/llm/test/inference_gpu/test_optimize_model.py

```diff
@@ -54,6 +58,126 @@ def test_optimize_model(Model, Tokenizer, model_path):
     diff = abs(logits_base_model - logits_optimized_model).flatten()
 
     assert any(diff) is False
+"""
+
+class Test_Optimize_Gpu_Model:
+    def setup(self):
+
+        self.layer_inputs = []
+        self.layer_outputs = []
+        self.pre_layer_outputs = []
+
+    def run_optimize_gpu_model(self, Model, Tokenizer, model_path, self_attn, prev_attn, lower_bound):
+        def forward_hook(module, input, output, layer_name):
+            self.layer_outputs.append(output)
+
+        def pre_forward_hook(module, input, output, layer_name):
+            self.pre_layer_outputs.append(output)
+
+        def new_forward_hook(module, input):
+            if model_path == os.environ.get('LLAMA2_7B_ORIGIN_PATH'):
+                repalcement_norm = model.model.norm
+                opt_model.model.norm = repalcement_norm
+
+        def pre_hook(module, input):
+            self.layer_inputs.append(input)
+
+        def replace_forward_hook(module, input, output, layer_name):
+            hidden_states, present_key_value = output
+            hidden_states = layer_states[0].detach()
```

Comment:
same as above, layer_states is not even defined at this point

Reply:
it's a hook and will only be executed with forward, when it was executed, there already has a list like layer_states definded.

Reply:
I know it can work, but it is awkward to write like this. In theory, the hook should be independent of what happens outside the hook?